aten/src/ATen/native/cpu/SoftMaxKernel.cpp

```diff
@@ -22,80 +22,118 @@
-        // NOLINTNEXTLINE(modernize-avoid-c-arrays,cppcoreguidelines-avoid-c-arrays)
-        scalar_t max_input_arr[CHUNK_SIZE];
-        for (int64_t ii = begin; ii < end; ii += CHUNK_SIZE) {
-          int64_t loop_end = CHUNK_SIZE;
-          if (ii + CHUNK_SIZE > end)
-            loop_end = end - ii;
-          for (const auto j : c10::irange(loop_end)) {
-            int64_t i = ii + j;
-            scalar_t* input_data = input_data_base + i * dim_size;
-            max_input_arr[j] = vec::reduce_all<scalar_t>(
-                [](Vec& x, Vec& y) { return vec::maximum(x, y); },
-                input_data,
-                dim_size);
-          }
-          for (const auto j : c10::irange(loop_end)) {
-            int64_t i = ii + j;
-            scalar_t* input_data = input_data_base + i * dim_size;
-            scalar_t max_input = max_input_arr[j];
-            tmp_sum_scalar[j] = vec::map_reduce_all<scalar_t>(
-                [max_input](Vec x) { return (x - Vec(max_input)).exp(); },
-                [](Vec x, Vec y) { return x + y; },
-                input_data,
-                dim_size);
-          }
-          // See [Note AVX-SSE transitions] for why this should call the
-          // vectorized version (aside from perf improvements).
-          vec::map(
-              [](Vec x) { return x.log(); },
-              tmp_sum_scalar,
-              tmp_sum_scalar,
-              loop_end);
-          for (const auto j : c10::irange(loop_end)) {
-            int64_t i = ii + j;
-            scalar_t* input_data = input_data_base + i * dim_size;
-            scalar_t* output_data = output_data_base + i * dim_size;
-            scalar_t tmp_sum = tmp_sum_scalar[j];
-            scalar_t max_input = max_input_arr[j];
-
-            // It's necessary to keep the order of the operations below.
-            // In some cases that input is large digits and the difference
-            // is small, if we compute `max_input` plus `tmp_sum` before,
-            // there would be a numerical problem. See an example in
-            // https://github.com/pytorch/pytorch/issues/11752#issuecomment-422883379
-            vec::map(
-                [tmp_sum, max_input](Vec x) { return x - Vec(max_input) - Vec(tmp_sum); },
-                output_data,
-                input_data,
-                dim_size);
-          }
-        }
-      });
+  parallel_for(0, outer_size, grain_size, [&](int64_t begin, int64_t end) {
+#ifndef _MSC_VER
+    // This snippet is not used with MSVC, because it leads to the build error
+    // 'expression did not evaluate to a constant'
+    std::unique_ptr<scalar_t[]> tmp_sum_scalar(new scalar_t[CHUNK_SIZE]);
+    std::unique_ptr<scalar_t[]> max_input_arr(new scalar_t[CHUNK_SIZE]);
+#else
+    // MSVC requires such a declaration of dynamic arrays
+    // Source: https://stackoverflow.com/a/33423538
+    scalar_t* tmp_sum_scalar = new scalar_t[CHUNK_SIZE];
+    scalar_t* max_input_arr = new scalar_t[CHUNK_SIZE];
+#endif
```

Comment:
Hi @jgong5, if you'd recommend, I'd remove the `unique_ptr` part for MSVC to make the code more readable. Thanks!